src/ripple/conditions/condition.h

```diff
@@ -0,0 +1,159 @@
+std::uint16_t constexpr condition_threshold_sha256 = 2;
+std::uint16_t constexpr condition_rsa_sha256       = 3;
+std::uint16_t constexpr condition_ed25519          = 4;
+
+// NIKB-TODO: These should be `enum class : std::uint32_t`
+std::uint32_t constexpr feature_sha256             = 1;
+std::uint32_t constexpr feature_preimage           = 2;
+std::uint32_t constexpr feature_prefix             = 4;
+std::uint32_t constexpr feature_threshold          = 8;
+std::uint32_t constexpr feature_rsa_pss            = 16;
+std::uint32_t constexpr feature_ed25519            = 32;
+
+struct condition_t
+{
+    std::uint16_t type;
+    std::uint16_t fulfillment_length;
+    std::uint32_t requires;
+    std::array<std::uint8_t, 32> fingerprint;
+
+    // Can this be deleted?
+    condition_t () = default;
+
+    condition_t (condition_t const&) = default;
+    condition_t (condition_t&&) = default;
+};
+
+inline
+bool
+operator== (condition_t const& lhs, condition_t const& rhs)
+{
+    return
+        lhs.type == rhs.type &&
+        lhs.requires == rhs.requires &&
+        lhs.fulfillment_length == rhs.fulfillment_length &&
+        lhs.fingerprint == rhs.fingerprint;
+}
+
+inline
+bool
+operator!= (condition_t const& lhs, condition_t const& rhs)
+{
+    return !(lhs == rhs);
+}
+
+/** Determine if a given condition is valid.
+
+    @note this function checks whether it understands the
+          type of the condition, and if so, whether it meets
+          the requirements mandated by the RFC.
+*/
+inline
+bool
+validate (condition_t const& c)
+{
+    // This check can never trigger because of the range of
+    // the fulfillment_length type. It's here in case the
+    // type is changed in the future.
+
+    // FIXME: maxFulfillmentLength
+    if (c.fulfillment_length > 65535)
+        return false;
+
+    if (c.type == condition_hashlock)
+        return (c.requires & (feature_sha256 | feature_preimage));
+
+    if (c.type == condition_prefix_sha256)
+        return (c.requires & (feature_sha256 | feature_prefix));
+
+    if (c.type == condition_rsa_sha256)
+        return (c.requires & (feature_rsa_pss | feature_sha256));
+
+    if (c.type == condition_ed25519)
+        return (c.requires & feature_ed25519);
+
+    return false;
+}
+
+/** True is type identifies cryptocondition in RFC.
+
+    @note: this function may return true even if the type
+           of condition isn't presently supported by this
+           implementation.
+*/
+inline
+bool
+isCondition (std::uint16_t v)
+{
+    switch(v)
+    {
+    case condition_hashlock:
+    case condition_prefix_sha256:
+    case condition_threshold_sha256:
+    case condition_rsa_sha256:
+    case condition_ed25519:
+        return true;
+
+    default:
+        return false;
+    }
+}
+
+/** Load a serialized condition either from its string or binary form */
+/** @{ */
+boost::optional<condition_t>
+load_condition(std::string s);
+
+boost::optional<condition_t>
+load_condition(Slice const& s);
```

Comment:
We pass `Slice` params by both value and `const&`. It's a nit, but I changed them all to pass by value as a minor cleanup. I'll include a commit with the changes in a later comment.